### PR TITLE
Add min/max user cases criteria to geospatial disbursement algorithm

### PR DESCRIPTION
**Pull request by Charl1996**
## Product Description
See demo video here: https://dimagi.slack.com/archives/C05EZGP724C/p1716369772880899

Two new criteria has been added for the disbursement algorithm on the geospatial settings page, namely `Minimum cases assigned per user` and `Maximum cases assigned per user`.

These criteria governs the min/max amount of cases that each user can have assigned to them by the algorithm. If the maximum number of cases per user * n_users is less than the total amount of assignable cases, the problem becomes infeasible, so we don't allow the user to do that and simply report an issue to the user:

image
(note that this banner is an update from the demo video; in the demo video the banner sits at the top of the page, but it makes more sense to sit at the top of the map)

## Technical Summary
[Ticket](https://dimagi.atlassian.net/browse/SC-3592)

The algorithm simply takes the configured `min_cases_per_user` and `max_cases_per_user` of the GeoConfig model into account when specifying the PuLP problem constraints see https://github.com/dimagi/commcare-hq/commit/33a5bc8844c8ad146b1f64ac6d7bf07e906c30c9.

## Feature Flag
Geospatial FF (`GEOSPATIAL`)

### Safety story
Tested locally and with unit tests

### Automated test coverage
Unit tests expanded

### QA Plan
QA ticket coming soon...

### Migrations
- [] The migrations in this code can be safely applied first independently of the code

The migration included needs the model definition in the code, so I don't think it can be applied independently?

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 1]
```diff
@@ -97,12 +99,20 @@ def get_decision_variables(x_dim, y_dim):
...
     @staticmethod
     def add_user_case_assignment_constraint(
-        lp_problem, decision_variables, user_count, case_count
+        lp_problem, decision_variables, user_count, case_count, min_cases=None, max_cases=None
     ):
-        # This constrain enforces cases being split equally between users.
-        max_constraint = int(case_count / user_count) + 1
+        # This constrain enforces the min/max amount of cases that could be assigned to each user,
+        # with the default being the cases split equally between users.
+        max_constraint = int(user_count / case_count) + 1
```
Comment (Charl1996) [inline corehq/apps/geospatial/routing_solvers/pulp.py]: For some reason I made a doozy and switched the `case_count` and `user_count` here (switched back [here](https://github.com/dimagi/commcare-hq/pull/34657/commits/91f5fd46be7283e096dc03e929be9680f9886efd))

File: corehq/apps/geospatial/migrations/0006_geoconfig_max_cases_per_user_and_more.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='geoconfig',
+            name='max_cases_per_user',
+            field=models.IntegerField(null=True),
+        ),
+        migrations.AddField(
+            model_name='geoconfig',
+            name='min_cases_per_user',
+            field=models.IntegerField(default=1),
```
Comment (mkangia) [inline corehq/apps/geospatial/migrations/0006_geoconfig_max_cases_per_user_and_more.py:21]: :+1:

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 3]
```diff
@@ -38,47 +38,91 @@ def solve(self, config, print_solution=False):
...
                                   f"user {self.user_locations[i]['id']}. "
                                   f"Cost: {costs[i][j]}")
-        else:
-            if print_solution:
-                print("No solution found.")
-        return None, solution
+            return None, solution
+
+        if print_solution:
+            print("No solution found.")
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/geospatial/routing_solvers/pulp.py:89]: or may be not?

**Comment by mkangia** [comment]
Looking good.

Happy to take a better review when ready for review.

File: corehq/apps/geospatial/routing_solvers/pulp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:89]: It was there when I got to the code, but I think I'll remove it. It's easy for a developer to just add it if someone needs it.

**Review by zandre-eng — APPROVED** [review]
Overall looks good. Only suggestion I would make is to maybe add some extra tests for the edge cases where there are no users or no cases. Also perhaps when there are more users than cases.

**Comment by Charl1996** [comment]
> Overall looks good. Only suggestion I would make is to maybe add some extra tests for the edge cases where there are no users or no cases. Also perhaps when there are more users than cases.

That's a good idea

**Review by zandre-eng — APPROVED** [review]
Thanks for the extra tests!

File: corehq/apps/geospatial/forms.py [hunk 4]
```diff
@@ -83,6 +85,22 @@ class Meta:
...
+        help_text=_("The algorithm which will be used to disburse cases between users"),
+    )
+    min_cases_per_user = forms.IntegerField(
+        label=_("Minimum cases assigned per user"),
+        help_text=_("The minimum number of cases each user can be assigned"),
+        required=True,
+        min_value=1,
+    )
+    max_cases_per_user_select = forms.BooleanField(
+        label=_("Specify a maximum number of cases per user"),
```
Comment (mkangia) [inline corehq/apps/geospatial/forms.py:97]: nit: `Specify maximum number `
(no 'a')

File: corehq/apps/geospatial/forms.py [hunk 5]
```diff
@@ -213,6 +244,11 @@ def clean(self):
         if algorithm == GeoConfig.ROAD_NETWORK_ALGORITHM and not token:
             raise ValidationError(_("Mapbox API token required"))
 
+        if not cleaned_data['max_cases_per_user_select']:
+            cleaned_data['max_cases_per_user'] = None
+        elif cleaned_data['max_cases_per_user'] < cleaned_data['min_cases_per_user']:
+            raise ValidationError(_("The maximum cases per user cannot be less than the minimum specified"))
```
Comment (mkangia) [inline corehq/apps/geospatial/forms.py:250]: :+1:

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 6]
```diff
@@ -33,52 +33,94 @@ def calculate_distance_matrix(self, config):
...
+    ):
+        # This constrain enforces the min/max amount of cases that could be assigned to each user,
+        # with the default being the cases split equally between users.
+        max_constraint = int(case_count / user_count) + 1
+        min_constraint = 0
+
+        if min_cases:
+            min_constraint = min_cases
+        if max_cases:
+            max_constraint = max_cases
```
Comment (mkangia) [inline corehq/apps/geospatial/routing_solvers/pulp.py:110]: nit: would be nice to make this explicit. It's more of a personal preference. I believe it is confusing when there is a default value set that is changed later in the function.

```
min_constraint = min_cases or 0
max_constraint = max_cases or int(case_count / user_count) + 1
```

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 7]
```diff
@@ -33,52 +33,94 @@ def calculate_distance_matrix(self, config):
...
+
+        for i in range(user_count):
+            lp_problem += pulp.lpSum([decision_variables[i, j] for j in range(case_count)]) <= max_constraint
+            lp_problem += pulp.lpSum([decision_variables[i, j] for j in range(case_count)]) >= min_constraint
```
Comment (mkangia) [inline corehq/apps/geospatial/routing_solvers/pulp.py:114]: I don't understand how this calculation works so I go with your judgement on a functional & performant solution here.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 8]
```diff
@@ -51,8 +52,26 @@ hqDefine("geospatial/js/geo_config", [
...
+
+        if (!data.max_cases_per_user) {
+            $("#div_id_max_cases_per_user").hide();
+        }
+
         self.plaintext_api_token = ko.observable(data.plaintext_api_token);
 
+        self.onSpecifyMaxCasesPerUser = function () {
+            var $maxCasesPerUserInput = $("#div_id_max_cases_per_user");
+            var $maxCasesPerUserCheckbox = $("#max_cases_per_user_checkbox_id");
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/geo_config.js:66]: nit: Is it a practice to use "_id" in the ID of HTML elements? We added div_id in the beginning. Good to check once about what is followed in HQ.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 9]
```diff
@@ -51,8 +52,26 @@ hqDefine("geospatial/js/geo_config", [
...
+
+            if ($maxCasesPerUserCheckbox.is(":checked")) {
+                $maxCasesPerUserInput.show();
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/geo_config.js:69]: nit: is it possible to do this via KO instead of jQuery? I believe such logic is more reliably put together with KO and you have a view model here already.

**Review by mkangia — APPROVED** [review]
Nice!

Left Jjust some non blocking suggestions.

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 10]
```diff
@@ -33,52 +33,94 @@ def calculate_distance_matrix(self, config):
 
         return distance_matrix
 
-    def solve(self, config, print_solution=False):
+    def solve(self, config):
         costs = self.calculate_distance_matrix(config)
+
+        if not costs:
+            # Infeasible solution
+            return None, None
```
Comment (ajeety4) [inline corehq/apps/geospatial/routing_solvers/pulp.py:41]: nit: maybe consider raising exceptions in case of infeasible solution or solution not found. Feel like could be more explicit.

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 11]
```diff
@@ -33,52 +33,94 @@ def calculate_distance_matrix(self, config):
 
         return distance_matrix
 
-    def solve(self, config, print_solution=False):
+    def solve(self, config):
```
Comment (ajeety4) [inline corehq/apps/geospatial/routing_solvers/pulp.py:36]: :+1:

**Review by ajeety4 — APPROVED** [review]
:+1: 
Just a non blocking comment

File: corehq/apps/geospatial/routing_solvers/pulp.py (hunk 10 above)
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:41]: I was initially thinking about raising an error, but then I thought it's not actually an error, it's just that there's no solution, hence returning a `None`.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 12]
```diff
@@ -55,23 +55,8 @@ hqDefine("geospatial/js/geo_config", [
         self.minCasesPerUser = ko.observable(data.min_cases_per_user);
         self.maxCasesPerUser = ko.observable(data.max_cases_per_user);
 
-        if (!data.max_cases_per_user) {
-            $("#div_id_max_cases_per_user").hide();
-        }
-
         self.plaintext_api_token = ko.observable(data.plaintext_api_token);
 
-        self.onSpecifyMaxCasesPerUser = function () {
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/geo_config.js]: This is nice.

**Review by ajeety4 — APPROVED** [review]

**Review by mkangia — APPROVED** [review]
